### PR TITLE
WIP Tokamak inspired fusion engine (Plasma Orientated Micro Fusion)

### PR DESCRIPTION
**Pull request by ninjanomnom**
**WIP**

ss 2017-04-14 at 07 19 06

The console UI and all numbers relating to balancing of reactions are temporary until I do more tests. The formulas however for dealing with those numbers are probably final unless changes are needed for something unforeseen. Final at least, for the prototype release.

And I must repeat that, this is for a prototype release where it barely works. It's the framework for everything and I plan on having multiple PRs flesh out how mods, power generation, maintenance, and containment failure work.

The map change isn't in this PR for now because I don't want to have to deal with conflicts because it's constantly being updated.

**Hydrogen** creates water vapor when burned but has a heat capacity larger than plasma (300)

**Water vapor** has been modified to not have any effect on temperature for now, might add a nerfed version back in.

**Fusion plasma** is a gas that degrades into hydrogen and plasma and increases the temperature when the temperature is below 12000 kelvin. It has a heat capacity of 600 but it's a bit tricky to handle normally for obvious reasons.

Added oxygen to scrubber filtering because I figured why not while adding hydrogen and water vapor to those same lists.

Injectors auto-connect to a nearby containment pipe on creation.

See [the google doc](https://docs.google.com/document/d/1ZKSywdmeT_ZpJKbokr_4_yYxg4iJcOj1JF-Eos_OzvE/edit?usp=sharing) for details on the planned machinery.

I also modified tgui/package.json as per #https://github.com/tgstation/tgstation/pull/25280 because I was having issues getting it to work. May as well have that in here.

TODO list for prototype:

- [x] Implement the gases needed
- [x] Implement the random items/materials needed
- [x] Finish core machinery
- [x] Finish management console
- [x] Fix major bugs from solo testing
- [ ] Get basic sprites

Pictures!

Injector sprite (by shuffl3)
pomfinjectorclean
Electromagnet sprite (by shuffl3)
electromagnet

:cl: ninjanomnom
add: Hydrogen with a heat capacity of 300!
add: A modified water vapor https://github.com/tgstation/tgstation/pull/20236/files (Thanks goof)
tweak: Oxygen is now an option for scrubbing.
wip: Plasma Orientated Micro Fusion engine [linktowikiguide]
wip: New salvaging and cargo goods: Engine upgrade modules
experiment: Random stats on some upgrade modules
imageadd: Sprites by people not me
/:cl:

### CODE REVIEW

**Comment by Qustinnus** [comment]
Post any pics you have please :D even the most basic stuff.

**Comment by ninjanomnom** [comment]
We have no pics, I have no sprites for any gases or canisters other than what already exists from the water vapor stuff, but you can look at goof's pr if you want to see a picture of that. I'll have an example picture of what the engine will look like once I get the sprites.

**Comment by Qustinnus** [comment]
I thought you had pics of the pipes and engine itself before :P my bad.

**Comment by ninjanomnom** [comment]
Thank shuffl3 for working hard on sprites while I dig myself deeper into atmos and pipenet code hell.

**Comment by ninjanomnom** [comment]
More git shenanigans ¯\_(ツ)_/¯

Most of the groundwork is in place now, biggest thing left to do is the console and it's tgui.

Almost all the code related to engine performance is in there.

Most the the device links are set up but I haven't been able to test things much yet.

The electromagnet and injector have temporary construction methods now.

**Comment by ninjanomnom** [comment]
The interactions between the core engine components are complete for the most part now. I probably missed something somewhere though. Currently no runtimes (from my code anyway). Largest thing left to do is the fusion console gui code. Engine upgrades need to be finished and added to station merchandise. Then I just need to make sure it actually works in practice.

**Comment by ninjanomnom** [comment]
@monster860 @Vivalas code is 95% done for a prototype, get me some reviews \o/

**Comment by PJB3005** [comment]
Ew those are some ugly sprites. 
tmp_3580-screenshot_20170417-125842-208190779

Still WiP.

**Comment by PJB3005** [comment]
tmp_3138-artwork_149243243530038828

File: code/modules/power/fusion/fusion_console.dm [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+/mob/camera/aiEye/remote/fusion_console/setLoc(var/T)
+	var/area/new_area = get_area(T)
+	if(new_area && new_area.name == "Engineering")
+		return ..()
+	else
+		return
+
+/obj/machinery/computer/camera_advanced/fusion_console
+	name = "Reaction Control Console"
+	desc = "Legend says if a non-engineer touches this they'll be found in maintenence with toolbox shaped bruises and memory loss"
```
Comment (monster860) [inline code/modules/power/fusion/fusion_console.dm]: kek

**Comment by PJB3005** [comment]
> no mention of Teslas or MeV.

pfff.

File: code/modules/power/fusion/fusion.dm [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+	if(istype(I,/obj/item/weapon/fusion_mod))
+		switch(add_part(I)) //Do bitflags maybe?
+			if(0)
+				user << "<span class='caution'>You install the mod.</span>"
+			if(1)
+				user << "<span class='caution'>This component has no slots for mods left!</span>"
+			if(2)
+				user << "<span class='caution'>This mod seems broken, you may want to reconstruct the engine component.</span>"
+			if(4)
+				user << "<span class='caution'>This mod is incompatible with an instaled mod.</span>"
```
Comment (monster860) [inline code/modules/power/fusion/fusion.dm]: > instaled

**Comment by ike709** [comment]
@PJB3005 I like those pipes

File: code/modules/power/fusion/fusion_console.dm (hunk 1 above)
Comment (ninjanomnom) [inline code/modules/power/fusion/fusion_console.dm]: ;)

**Comment by ninjanomnom** [comment]
@monster860 fixed the stuff that can be fixed. Would you mind taking a look at the injector btw? I can't seem to get the atmos terminal to accept hydrogen gas for some reason even though ingame all the dirs are correct.

**Comment by ninjanomnom** [comment]
I believe I found the issue. I should have figured that just changing the dir on a pipe isnt enough to make it reconnect to new pipes. As a result I've given /obj/machinery/atmospherics/components/unary/terminal a new change_dir proc to make it easier and handle all that.

**Comment by ninjanomnom** [comment]
I'll re-open this when I have time to work on it again after bug fixing the rebase and atomizing the upgrade system I need for this.

**Pull request closed without merging**